src/flag_gems/ops/exponential_.py

```diff
@@ -0,0 +1,99 @@
+import logging
+
+import torch
+import triton
+import triton.language as tl
+
+from flag_gems.utils.random_utils import philox_cuda_seed_offset, uint_to_uniform_float
+
+
+def heur_block(args):
+    if args["N"] <= 512:
+        return 512
+    else:
+        return 1024
+
+
+def heur_num_warps(args):
+    if args["N"] <= 512:
+        return 4
+    elif args["N"] <= 1024:
+        return 8
+    else:
+        return 16
+
+
+@triton.heuristics(
+    {
+        "BLOCK": heur_block,
+        "num_warps": heur_num_warps,
+    }
+)
+@triton.jit(do_not_specialize=["philox_seed", "philox_offset"])
+def fused_exponential_kernel(
+    out_ptr,
+    N,
+    lambd,
+    eps,
+    philox_seed,
+    philox_offset,
+    BLOCK: tl.constexpr,
+):
+    UNROLL: tl.constexpr = 4
+    philox_seed = philox_seed.to(tl.int64)
+    philox_offset = philox_offset.to(tl.int64)
+    c0 = (philox_offset & 0xFFFFFFFF).to(tl.uint32)
+    c1 = ((philox_offset >> 32) & 0xFFFFFFFF).to(tl.uint32)
+    i4 = tl.program_id(0) * BLOCK + tl.arange(0, BLOCK)
+    c0 += i4
+    _O = c0 * 0
+    r0, r1, r2, r3 = tl.philox(philox_seed, c0, c1, _O, _O)
+    r0 = uint_to_uniform_float(r0)
+    r1 = uint_to_uniform_float(r1)
+    r2 = uint_to_uniform_float(r2)
+    r3 = uint_to_uniform_float(r3)
+    off_0 = tl.program_id(0) * BLOCK * UNROLL + tl.arange(0, BLOCK)
+    off_1 = off_0 + BLOCK
+    off_2 = off_1 + BLOCK
+    off_3 = off_2 + BLOCK
+    r0 = transform_exponential(r0, lambd, eps)
+    r0 = transform_exponential(r1, lambd, eps)
+    r0 = transform_exponential(r2, lambd, eps)
+    r0 = transform_exponential(r3, lambd, eps)
+    tl.store(out_ptr + off_0, r0, mask=off_0 < N, eviction_policy="evict_first")
+    tl.store(out_ptr + off_1, r1, mask=off_1 < N, eviction_policy="evict_first")
+    tl.store(out_ptr + off_2, r2, mask=off_2 < N, eviction_policy="evict_first")
+    tl.store(out_ptr + off_3, r3, mask=off_3 < N, eviction_policy="evict_first")
+
+
+@triton.jit
+def transform_exponential(u, lambd, eps):
+    eps1 = -0.5 * eps
+    is_subnormal = u > 1.0 + eps1
+    log = tl.where(is_subnormal, eps1, tl.math.log(u))
+    v = -1.0 / lambd * log
+    return v
```

Comment:
The inverse cdf  of exponential distribution is

`x = - log(1-y) / lambda`

y in [0, 1].

First sample y from a Uniform distribution U(0,1) then apply the inverse cdf.

Does this implementation exploits the feature that `1-y` is also a sample from `U(0,1)` to eliminate a subtraction? Then change the formula to 

`x = - log(y) / lambda`?

Reply:
I got it. Uniform is of the same distribution as 1-u.

Reply:
but what is eps used for? and what't the expected data type of is_subnormal?